lambda/translprim.ml

```diff
@@ -471,14 +449,68 @@ let simplify_constant_constructor = function
 *)
 let glb_array_type t1 t2 =
   match t1, t2 with
+  (* No GLB; only used in the [Obj.magic] case *)
   | Pfloatarray, (Paddrarray | Pintarray)
   | (Paddrarray | Pintarray), Pfloatarray -> t1
 
+  (* Compute the correct GLB *)
   | Pgenarray, x | x, Pgenarray -> x
   | Paddrarray, x | x, Paddrarray -> x
   | Pintarray, Pintarray -> Pintarray
   | Pfloatarray, Pfloatarray -> Pfloatarray
 
+let glb_array_ref_type t1 t2 =
+  match t1, t2 with
+  (* No GLB; only used in the [Obj.magic] case *)
+  | Pfloatarray_ref _, (Paddrarray | Pintarray)
+  | (Paddrarray_ref | Pintarray_ref), Pfloatarray -> t1
+
+  (* Compute the correct GLB *)
+
+  (* Pgenarray >= _ *)
+  | (Pgenarray_ref _ as x), Pgenarray -> x
+  | Pgenarray_ref _, Pintarray -> Pintarray_ref
+  | Pgenarray_ref _, Paddrarray -> Paddrarray_ref
+  | Pgenarray_ref mode, Pfloatarray -> Pfloatarray_ref mode
+  | (Paddrarray_ref | Pintarray_ref | Pfloatarray_ref _ as x), Pgenarray -> x
```

Comment:
To better match the code in `glb_array_type`, maybe fold the last case into the first one.

Reply:
(same in `glb_array_set_type` below)

Reply:
I don't understand this comment -- in `glb_array_type`, things are separated out exactly the same way as I have it, with `float, addr | int` and `addr | int, float` cases leading things off.

Reply:
@mshinwell What's the verdict here?

Reply:
OK, let's leave this then.